src/sqlfluff/core/parser/grammar/base.py

```diff
@@ -475,9 +488,15 @@ def _bracket_sensitive_look_ahead_match(cls, segments, matchers, parse_context):
                     )
 
                     if match:
-                        if matcher in start_brackets:
+                        # NB: We can only consider this as a nested bracket if the start
+                        # and end tokens are not the same. If a matcher is both a start and
+                        # end token we cannot deepen the bracket stack.
+                        if matcher in start_brackets and matcher not in end_brackets:

```

Comment:
@alanmcruickshank: I think you contributed this part of the code. Possibly you were thinking about things like single or double quotes? IIUC, those are handled in a separate part of the parser, and they delineate a single token with "stuff" inside. So... I think I agree with you, but I don't know the parser that well. Do you think this could (should?) be removed? Or I could add a code comment something like this confusing one. 😄

Reply:
For now, I'm planning to add this comment:
```
                        # NB: We can only consider this as a nested bracket if the start
                        # and end tokens are not the same. If a matcher is both a start and
                        # end token we cannot deepen the bracket stack. In general, quoted
                        # strings are a typical example where the start and end tokens are
                        # the same. Currently, though, quoted strings are handled elsewhere
                        # in the parser, and there are no cases where *this* code has to
                        # handle identical start and end brackets. For now, consider this
                        # a small, speculative investment in a possible future requirement.
```